### PR TITLE
tests: reconcile

### PR DESCRIPTION
**Pull request by pkliczewski**
This PR provides tests for reconcile logic and introduces ability to mock oVirt client.

Signed-off-by: Piotr Kliczewski <piotr.kliczewski@gmail.com>

### CODE REVIEW

File: pkg/providers/provider.go [hunk 1]
```diff
@@ -41,3 +43,21 @@ type Mapper interface {
...
+// SourceClientFactory provides default client factory implementation
+type SourceClientFactory struct{}
+
+// NewSourceClientFactory creates new factory
+func NewSourceClientFactory() *SourceClientFactory {
+	return &SourceClientFactory{}
+}
+
+// NewOvirtClient creates new Ovirt clients
+func (f *SourceClientFactory) NewOvirtClient(dataMap map[string]string) (aclient.VMClient, error) {
```
Comment (jakub-dzon) [inline pkg/providers/provider.go]: I think it should not be located here - that package should be general and shouldn't know anything about its implementations.

File: pkg/client/client.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package client
+
+// Facotry creates new clients
+type Facotry interface {
+	NewOvirtClient(dataMap map[string]string) (VMClient, error)
+}
+
+// VMClient provides interface how source virtual machines should be fetched
+type VMClient interface {
```
Comment (jakub-dzon) [inline pkg/client/client.go:9]: I am not sure that it needs to be so generic - other providers might need other operations and the operations are executed only by the concrete providers.

What do you think about creating provider factory, that would assemble providers with what they need instead? Is control over vm client needed in "general" packages? All that can be mocked on the client level can be mocked on provider (from controller perspective). I think vm client should be an implementation detail of each provider.

Generally I think that layer of code should only have access to abstration over its lower level if it needs to be both extensible and loosely-coupled. In this case we give controller access to abstraction over code two levels below.

**Review by jakub-dzon — CHANGES_REQUESTED** [review]

File: pkg/client/client.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package client
+
+// Facotry creates new clients
+type Facotry interface {
```
Comment (machacekondra) [inline pkg/client/client.go]: Factory

Comment (pkliczewski) [reply pkg/client/client.go]: will fix.

File: pkg/providers/provider.go (hunk 1 above)
Comment (pkliczewski) [reply pkg/providers/provider.go]: Do you suggest to move it to `OvirtProvider`?

File: pkg/client/client.go (hunk 2 above)
Comment (pkliczewski) [reply pkg/client/client.go:9]: Interesting idea, will explore.

File: pkg/providers/provider.go (hunk 1 above)
Comment (jakub-dzon) [reply pkg/providers/provider.go]: Yes, or to the potential provider factory.

File: pkg/client/client.go (hunk 2 above)
Comment (pkliczewski) [reply pkg/client/client.go:9]: I thought about it how to do it well and it seems that provider factory won't bring a lot of value. We still would need a way to detach client and provider life cycles. We want to use the client when it is needed as specified in issue https://github.com/kubevirt/vm-import-operator/issues/64. I will move it to `OvirtProvider`.

Comment (jakub-dzon) [reply pkg/client/client.go:9]: Do you plan to use it directly from the controller? Won't provider lifecycle methods (maybe split/refactored) be sufficient?

Comment (pkliczewski) [reply pkg/client/client.go:9]: The idea is to have a factory which we can use to create client. In a unit test we need to have a way to mock the client. Provider methods are not sufficient for it.

Comment (jakub-dzon) [reply pkg/client/client.go:9]: :+1:

**Comment by pkliczewski** [comment]
@jakub-dzon Please review again

**Comment by jakub-dzon** [comment]
I don't understand why control over ovirt client creation is needed in the controller tests - is the controller going to interact with it directly as a solution to #64? If not, then why not just mock the provider and test ovirt client-related edge cases in ovirt provider tests?

**Comment by pkliczewski** [comment]
The idea is that we want to decouple client and provider life cycle. We do not want to connect when the provider is created and only connect when needed. We could come up with provider factory and mock it like we do for there tests but preparation to fix https://github.com/kubevirt/vm-import-operator/issues/64 needs to be done anyway. This PR helps to solve one problem which helps to fix the other one.

**Review by jakub-dzon — APPROVED** [review]

**Review by machacekondra — APPROVED** [review]

**Comment by machacekondra** [comment]
@pkliczewski Can you rebase? I will merge then.

File: pkg/controller/virtualmachineimport/virtualmachineimport_controller.go [hunk 4]
```diff
@@ -7,6 +7,7 @@ import (
 	"time"
 
 	v2vv1alpha1 "github.com/kubevirt/vm-import-operator/pkg/apis/v2v/v1alpha1"
+	aclient "github.com/kubevirt/vm-import-operator/pkg/client"
```
Comment (masayag) [inline pkg/controller/virtualmachineimport/virtualmachineimport_controller.go:10]: just a nit - I'd call it pclient (p for provider), but we can leave it as is

**Comment by pkliczewski** [comment]
@machacekondra please take a look

**Review by machacekondra — APPROVED** [review]

**Review by masayag — APPROVED** [review]

File: pkg/controller/virtualmachineimport/virtualmachineimport_controller.go (hunk 4 above)
Comment (pkliczewski) [reply pkg/controller/virtualmachineimport/virtualmachineimport_controller.go:10]: Naming is the hardest thing for me. Will push a PR to change it. Thanks!
